File: bigdoors_bp/scripts/domain/RotationMath.js
/**
 * Pure rotation math for door panels.
 *
 * Domain layer — no @minecraft/server imports.
 *
 * Minecraft Bedrock coordinates: +X east, +Z south, +Y up.
 * Visual clockwise from above: N → E → S → W.
 *
 * Formula A (visual CW from above):
 *   newX = hinge.x - (oldZ - hinge.z)
 *   newZ = hinge.z + (oldX - hinge.x)
 *
 * Formula B (visual CCW from above):
 *   newX = hinge.x + (oldZ - hinge.z)
 *   newZ = hinge.z - (oldX - hinge.x)
 */

/**
 * Rotate a position 90 degrees clockwise (viewed from above) around the hinge.
 * Horizontal mode: rotates in the X/Z plane, Y unchanged.
 */
export function rotateCW(pos, hingePos) {
  const dx = pos.x - hingePos.x;
  const dz = pos.z - hingePos.z;
  return {
    x: Math.round(hingePos.x - dz),
    y: pos.y,
    z: Math.round(hingePos.z + dx),
  };
}

/**
 * Rotate a position 90 degrees counter-clockwise (viewed from above) around the hinge.
 * Horizontal mode: rotates in the X/Z plane, Y unchanged.
 */
export function rotateCCW(pos, hingePos) {
  const dx = pos.x - hingePos.x;
  const dz = pos.z - hingePos.z;
  return {
    x: Math.round(hingePos.x + dz),
    y: pos.y,
    z: Math.round(hingePos.z - dx),
  };
}

/**
 * Rotate a position 90 degrees clockwise in vertical mode around a horizontal axis.
 * The rotation axis is perpendicular to the facing direction:
 * - Facing north/south: axis runs E/W → rotate in Y/Z plane (Y and Z change, X unchanged).
 * - Facing east/west: axis runs N/S → rotate in Y/X plane (Y and X change, Z unchanged).
 *
 * @param {string} facing 'north'|'south'|'east'|'west'
 */
export function rotateVerticalCW(pos, hingePos, facing) {
  if (facing === "north" || facing === "south") {
    const dz = pos.z - hingePos.z;
    const dy = pos.y - hingePos.y;
    return {
      x: pos.x,
      y: Math.round(hingePos.y - dz),
      z: Math.round(hingePos.z + dy),
    };
  }
  // east or west — rotate in Y/X plane
  const dx = pos.x - hingePos.x;
  const dy = pos.y - hingePos.y;
  return {
    x: Math.round(hingePos.x + dy),
    y: Math.round(hingePos.y - dx),
    z: pos.z,
  };
}

/**
 * Rotate a position 90 degrees counter-clockwise in vertical mode.
 */
export function rotateVerticalCCW(pos, hingePos, facing) {
  if (facing === "north" || facing === "south") {
    const dz = pos.z - hingePos.z;
    const dy = pos.y - hingePos.y;
    return {
      x: pos.x,
      y: Math.round(hingePos.y + dz),
      z: Math.round(hingePos.z - dy),
    };
  }
  const dx = pos.x - hingePos.x;
  const dy = pos.y - hingePos.y;
  return {
    x: Math.round(hingePos.x - dy),
    y: Math.round(hingePos.y + dx),
    z: pos.z,
  };
}

/**
 * Returns the appropriate rotation function for the given mode/facing/direction.
 * The returned function has signature (pos, hingePos) => pos.
 */
export function getRotateFn(mode, facing, direction) {
  if (mode === "vertical") {
    const vertFn = direction === "cw" ? rotateVerticalCW : rotateVerticalCCW;
    return (pos, hingePos) => vertFn(pos, hingePos, facing);
  }
  return direction === "cw" ? rotateCW : rotateCCW;
}

/**
 * Compute intermediate arc positions between current and rotated position
 * for entity sweep purposes. Returns an array of {x,y,z} positions the
 * panel passes through during the 90-degree sweep.
 *
 * @param {{x:number,y:number,z:number}} pos - panel position
 * @param {{x:number,y:number,z:number}} hingePos - hinge center
 * @param {'cw'|'ccw'} [direction='cw'] - rotation direction
 * @param {string} [mode='horizontal'] - 'horizontal' or 'vertical'
 * @param {string} [facing=''] - hinge facing for vertical mode
 */
export function computeArcPositions(pos, hingePos, direction = "cw", mode = "horizontal", facing = "") {
  const cos45 = Math.SQRT1_2;
  const sin45 = Math.SQRT1_2;
  const sign = direction === "ccw" ? -1 : 1;

  if (mode === "vertical") {
    if (facing === "north" || facing === "south") {
      const dz = pos.z - hingePos.z;
      const dy = pos.y - hingePos.y;
      return [{
        x: pos.x,
        y: Math.round(hingePos.y + dy * cos45 - sign * dz * sin45),
        z: Math.round(hingePos.z + sign * dy * sin45 + dz * cos45),
      }];
    }
    const dx = pos.x - hingePos.x;
    const dy = pos.y - hingePos.y;
    return [{
      x: Math.round(hingePos.x + dx * cos45 + sign * dy * sin45),
      y: Math.round(hingePos.y - sign * dx * sin45 + dy * cos45),
      z: pos.z,
    }];
  }

  const dx = pos.x - hingePos.x;
  const dz = pos.z - hingePos.z;
  const midX = Math.round(hingePos.x + dx * cos45 - sign * dz * sin45);
  const midZ = Math.round(hingePos.z + sign * dx * sin45 + dz * cos45);
  return [{ x: midX, y: pos.y, z: midZ }];
}

File: java-bigdoors/src/main/java/com/bigdoors/domain/RotationMath.java
package com.bigdoors.domain;

import com.bigdoors.util.BlockPos3;

import java.util.ArrayList;
import java.util.List;
import java.util.function.BiFunction;

/**
 * Pure rotation math for door panel positions around a hinge.
 * No Minecraft imports allowed.
 */
public final class RotationMath {

    private RotationMath() {}

    /**
     * Rotate 90 degrees clockwise (looking down, Y-axis).
     * CW: newX = hinge.x - (oldZ - hinge.z), newZ = hinge.z + (oldX - hinge.x)
     */
    public static BlockPos3 rotateCW(BlockPos3 pos, BlockPos3 hingePos) {
        int dx = pos.x() - hingePos.x();
        int dz = pos.z() - hingePos.z();
        return new BlockPos3(
                hingePos.x() - dz,
                pos.y(),
                hingePos.z() + dx
        );
    }

    /**
     * Rotate 90 degrees counter-clockwise (looking down, Y-axis).
     * CCW: newX = hinge.x + (oldZ - hinge.z), newZ = hinge.z - (oldX - hinge.x)
     */
    public static BlockPos3 rotateCCW(BlockPos3 pos, BlockPos3 hingePos) {
        int dx = pos.x() - hingePos.x();
        int dz = pos.z() - hingePos.z();
        return new BlockPos3(
                hingePos.x() + dz,
                pos.y(),
                hingePos.z() - dx
        );
    }

    /**
     * Rotate 90 degrees clockwise in the vertical plane defined by facing direction.
     */
    public static BlockPos3 rotateVerticalCW(BlockPos3 pos, BlockPos3 hingePos, String facing) {
        if ("north".equals(facing) || "south".equals(facing)) {
            int dx = pos.x() - hingePos.x();
            int dy = pos.y() - hingePos.y();
            return new BlockPos3(
                    hingePos.x() + dy,
                    hingePos.y() - dx,
                    pos.z()
            );
        } else {
            // east or west
            int dz = pos.z() - hingePos.z();
            int dy = pos.y() - hingePos.y();
            return new BlockPos3(
                    pos.x(),
                    hingePos.y() - dz,
                    hingePos.z() + dy
            );
        }
    }

    /**
     * Rotate 90 degrees counter-clockwise in the vertical plane defined by facing direction.
     */
    public static BlockPos3 rotateVerticalCCW(BlockPos3 pos, BlockPos3 hingePos, String facing) {
        if ("north".equals(facing) || "south".equals(facing)) {
            int dx = pos.x() - hingePos.x();
            int dy = pos.y() - hingePos.y();
            return new BlockPos3(
                    hingePos.x() - dy,
                    hingePos.y() + dx,
                    pos.z()
            );
        } else {
            // east or west
            int dz = pos.z() - hingePos.z();
            int dy = pos.y() - hingePos.y();
            return new BlockPos3(
                    pos.x(),
                    hingePos.y() + dz,
                    hingePos.z() - dy
            );
        }
    }

    /**
     * Returns the appropriate rotation function for the given mode, facing, and direction.
     */
    public static BiFunction<BlockPos3, BlockPos3, BlockPos3> getRotateFn(String mode, String facing, String direction) {
        if ("vertical".equals(mode)) {
            if ("cw".equals(direction)) {
                return (pos, hinge) -> rotateVerticalCW(pos, hinge, facing);
            } else {
                return (pos, hinge) -> rotateVerticalCCW(pos, hinge, facing);
            }
        }
        if ("cw".equals(direction)) {
            return RotationMath::rotateCW;
        }
        return RotationMath::rotateCCW;
    }

    /**
     * Compute the arc positions for a 90-degree CW rotation: the 45-degree midpoint
     * and the final 90-degree destination.
     */
    public static List<BlockPos3> computeArcPositions(BlockPos3 pos, BlockPos3 hingePos) {
        double dx = pos.x() - hingePos.x();
        double dz = pos.z() - hingePos.z();
        double cos45 = Math.sqrt(0.5); // ≈ 0.7071

        // 45-degree midpoint
        double midX = hingePos.x() + dx * cos45 - dz * cos45;
        double midZ = hingePos.z() + dx * cos45 + dz * cos45;
        BlockPos3 midpoint = new BlockPos3(
                (int) Math.round(midX),
                pos.y(),
                (int) Math.round(midZ)
        );

        // 90-degree destination (same as rotateCW)
        BlockPos3 destination = rotateCW(pos, hingePos);

        List<BlockPos3> result = new ArrayList<>(2);
        result.add(midpoint);
        result.add(destination);
        return result;
    }
}
